packages/amesos2/src/Amesos2_STRUMPACK_def.hpp

```diff
@@ -0,0 +1,519 @@
+//
+//           Amesos2: Templated Direct Sparse Solver Package
+//                  Copyright 2011 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+
+/**
+   \file   Amesos2_STRUMPACK_def.hpp
+   \author Pieter Ghysels <pghysels@lbl.gov>, Lisa Claus <lclaus@lbl.gov>
+   \date   Wed April 8 2020
+
+   \brief  Definitions for the Amesos2 STRUMPACK solver interface
+*/
+
+#ifndef AMESOS2_STRUMPACK_DEF_HPP
+#define AMESOS2_STRUMPACK_DEF_HPP
+
+#include <Teuchos_Tuple.hpp>
+#include <Teuchos_StandardParameterEntryValidators.hpp>
+#include <Teuchos_DefaultMpiComm.hpp>
+
+#include "Amesos2_SolverCore_def.hpp"
+#include "Amesos2_Util.hpp"
+
+#include <memory>
+#include "StrumpackSparseSolverMPIDist.hpp"
+
+namespace Amesos2 {
+
+
+  template <class Matrix, class Vector>
+  STRUMPACK<Matrix,Vector>::STRUMPACK(Teuchos::RCP<const Matrix> A,
+                                      Teuchos::RCP<Vector> X,
+                                      Teuchos::RCP<const Vector> B)
+    : SolverCore<Amesos2::STRUMPACK,Matrix,Vector>(A, X, B)
+
+  {
+    using Teuchos::Comm;
+    // It's OK to depend on MpiComm explicitly here, because
+    // STRUMPACK requires MPI anyway.
+    using Teuchos::MpiComm;
```

Comment:
I don't think Amesos2 knows STRUMPACK will depend on MPI. This file will result in a compilation error if someone turns off MPI in a build with STRUMPACK. This might need an ifdef check for MPI

Reply:
We'll add this check for MPI.
STRUMPACK can also be compiled without MPI, so we can add a slightly different interface in the #else clause that does not rely on MPI.

Reply:
> > Thanks ! Looks good to me. Please see below for some simple comments. BTW, it looks like you are doing ROOTED and the redistributing the data back in the format STRUMPACK wants. Can STRUMPACK take 1D row distributed matrix ?
> 
> We are a bit confused about the meaning of the ROOTED parameter. Does this mean the matrix is gathered on the root process?
> STRUMPACK does take the sparse matrix using a 1D block row compressed sparse row distribution. Should we use `Amesos2::EDistribution::DISTRIBUTED` or `Amesos2::EDistribution::DISTRIBUTED_NO_OVERLAP` instead of ROOTED?

Please see
https://github.com/trilinos/Trilinos/blob/69f8850c25bdead51b38866ae34d4fa40947b67f/packages/amesos2/src/Amesos2_TypeDecl.hpp#L123

You would use Distributed.